[SPIR-V] Add pass to remove spv_ptrcast intrinsics

Keenuts opened the pull request:
OpenCL is allowed to cast pointers, meaning they can resolve some type mismatches this way. In logical SPIR-V, those are restricted. This new pass legalizes such pointer cast when targeting logical SPIR-V.

For now, this pass supports 3 cases we witnessed:
 - loading a vec3 from a vec4*.
 - loading a scalar from a vec*.
 - loading the 1st element of an array.

Keenuts committed:
[SPIR-V] Add pass to remove spv_ptrcast intrinsics

OpenCL is allowed to cast pointers, meaning they can resolve some
type mismatches this way. In logical SPIR-V, those are restricted.
This new pass legalizes such pointer cast when targeting logical SPIR-V.

For now, this pass supports 3 cases we witnessed:
 - loading a vec3 from a vec4*.
 - loading a scalar from a vec*.
 - loading the 1st element of an array.

VyacheslavLevytskyy commented on llvm/lib/Target/SPIRV/SPIRVEmitIntrinsics.cpp:
```diff
@@ -142,23 +136,10 @@ class SPIRVEmitIntrinsics
   Type *reconstructType(Value *Op, bool UnknownElemTypeI8,
                         bool IsPostprocessing);
 
   void buildAssignType(IRBuilder<> &B, Type *ElemTy, Value *Arg);
```
Probably it'd be better to keep buildAssignType close to buildAssignPtr, even if all current users are within EmitIntrinsics.

VyacheslavLevytskyy commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -0,0 +1,265 @@
+
+  // Replace all uses of a |Old| with |New| updates the global registry type
+  // mappings.
+  void replaceAllUsesWith(Value *Old, Value *New) {
```
If we can move this to GlobalRegistry, it becomes a part of its tiny-api for assistance in type inference, and things become more structured, because we can then avoid code duplication in SPIRVEmitIntrinsics::replaceAllUsesWith and call this smaller version of GlobalRegistry::replaceAllUsesWith(Value *Old, Value *New) both from SPIRVEmitIntrinsics::replaceAllUsesWith and SPIRVEmitIntrinsics::transformLoad. What you think?

Keenuts replied:
Yes, fine with this, I'll move this to GR

VyacheslavLevytskyy commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -0,0 +1,265 @@
+
+  // Builds the `spv_assign_type` assigning |Ty| to |Value| at the current
+  // builder position.
+  void buildAssignType(IRBuilder<> &B, Type *Ty, Value *Arg) {
```
The same thought here, it looks like keeping all type inference related staff in one place (currently it's going to be GlobalRegistry) instead of dispersing over different modules may help in future.

I've been thinking about middle to long-term plans to reorganize type inference to a separate pass, splitting emit-intrinsics, and making type inference slightly more universal, so this is probably not the last anticipated refactoring and having that in mind a bigger picture is that we should somehow try factoring out and grouping together type inference related functions.

Keenuts replied:
I also think it would be beneficial to have a first pass which only emits type assigment intrinsics, while keeping a classic IR, and then, a late pass after legalization to generate the spv.load/spv.gep etc.
(Working on this pass was a bit painful because of this weird state where some valid IR instructions were not usable because not expected)

s-perron commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -0,0 +1,265 @@
+  // Loads parts of the vector of type |From| from the pointer |Source| and
+  // create a new vector of type |To|. |To| must be a vector type, and element
+  // types of |To| and |From| must match. Returns the loaded value.
+  Value *loadVectorFromVector(IRBuilder<> &B, FixedVectorType *From,
+                              FixedVectorType *To, Value *Source) {
```
The variable names could use more info. The different between `From` and `Source` is not immediately obvious. I have to go back and look at the defintions to see one is a type, and the other a value.

```suggestion
  Value *loadVectorFromVector(IRBuilder<> &B, FixedVectorType *SourceType,
                              FixedVectorType *TargetType, Value *Source) {
```

s-perron commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -0,0 +1,265 @@
+    LoadInst *NewLoad = B.CreateLoad(From, Source);
+    buildAssignType(B, From, NewLoad);
+
+    auto ConstInt = ConstantInt::get(IntegerType::get(B.getContext(), 32), 0);
```
```suggestion
    auto Zero = ConstantInt::get(IntegerType::get(B.getContext(), 32), 0);
```

s-perron commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -0,0 +1,265 @@
+        SmallVector<Value *, 3> Args = {Output, Extracted, B.getInt64(I)};
+        SmallVector<Type *, 4> Types = {Args[0]->getType(), Args[0]->getType(),
+                                        Args[1]->getType(), Args[2]->getType()};
+        Output = B.CreateIntrinsic(Intrinsic::spv_insertelt, {Types}, {Args});
```
For the future, the best way to do this is to use a single OpVectorShuffle. That intrinsic does not seem to exist yet, so no need to do it now, but we should have a TODO or something tagged with an issue. Is that how LLVM does things?

s-perron replied:
Can you use https://llvm.org/doxygen/classllvm_1_1ShuffleVectorInst.html?

Keenuts replied:
I misread the OpVectorShuffle/shuffle instruction and thought the 2 vectors length had to match.. there is an LLVM instruction `shufflevector`, maybe it's already lowered into `OpVectorShuffle`, I'll check.

Keenuts replied:
Changed the lowering to use OpVectorShuffle, way way better. Thanks!

s-perron commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -0,0 +1,265 @@
+
+  // Loads a single scalar of type |To| from the vector pointed by |Source| of
+  // the type |From|. Returns the loaded value.
+  Value *loadScalarFromVector(IRBuilder<> &B, Value *Source,
```
Why do we want to load the whole vector? We could do an access chain on the vector, and loads a single value. Does `spv_gep` not work on vectors?

If it does, we could try to reuse `loadFirstValueFromArray` (maybe rename to `loadFromValueFromAggregate`) for both cases.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -0,0 +1,265 @@
+  // Loads the first value in an array pointed by |Source| of type |From|. Load
+  // flags will be copied from |BadLoad|, which should be the illegal load being
+  // legalized. Returns the loaded value.
+  Value *loadFirstValueFromArray(IRBuilder<> &B, ArrayType *From, Value *Source,
```
All you use `From` for is to get the element type. You could pass in the element type directly, and to make this more general. It could work for vectors, structs, and arrays.

s-perron reviewed (commented):
In general, this looks good. I just have some minor tweaks. The main thing is that the focus of the pass should be to remove pointer casts. The loads themselves are fine. We may also need to fix stores, and possibly other types of uses. We should rename the pass to reflect that.

Keenuts committed:
Update llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp

Co-authored-by: Steven Perron <stevenperron@google.com>

Keenuts commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -0,0 +1,265 @@
+  void transformLoad(IRBuilder<> &B, LoadInst *LI, Value *CastedOperand,
+                     Value *OriginalOperand) {
+    Type *FromTy = GR->findDeducedElementType(OriginalOperand);
+    Type *ToTy /*-fruity*/ = GR->findDeducedElementType(CastedOperand);
```
Bad word-play tutti-frutty to myself I forgot to remove 🤦‍♂

Keenuts commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -0,0 +1,265 @@
+      Users.push_back(U.getUser());
+
+    for (Value *User : Users) {
+      if (LoadInst *LI = dyn_cast<LoadInst>(User)) {
+        transformLoad(B, LI, CastedOperand, OriginalOperand);
+        continue;
+      }
```
Stores would also be required yes. I'll rename the pass.

s-perron commented on llvm/lib/Target/SPIRV/SPIRVLegalizePointerLoad.cpp:
```diff
@@ -13,8 +13,8 @@
 // This pass relies on the assign_ptr_type intrinsic to deduce the type of the
-// pointed values, must replace all occurences of `ptrcast`. This is why
-// unhandled cases are reported as unreachable: we MUST cover all cases.
+// pointee. All occurrences of `ptrcast` must be replaced because the lead to
```
```suggestion
// pointee. All occurrences of `ptrcast` must be replaced because they lead to
```

Keenuts committed:
pr-feedback: move to GR & alternative lowering

Moved more EmitIntrinsic function to GR, and changed the lowering to use
shuffle and reuse array load logic.

Keenuts commented:
Thanks for the review! Addresses all the comments
